libs/libcommon/include/common/avx2_memcpy.h

```diff
@@ -0,0 +1,162 @@
+    const char * __restrict src = reinterpret_cast<const char * __restrict>(src_);
+
+    void * ret = dst;
+
+#define MEM_CP_END(n) tiflash_compiler_builtin_memcpy(dst + size - (n), src + size - (n), (n));
+#define MEM_CP_HEAD(n) tiflash_compiler_builtin_memcpy(dst, src, (n));
+#ifdef AVX2_MEMCPY_USE_PREFETCH
+#define PREFETCH(addr) __builtin_prefetch(addr)
+#else
+#define PREFETCH(addr)
+#endif
+
+    constexpr int block32_size = details::BLOCK32_SIZE;
+
+    if (likely(size <= block32_size))
+    {
+        if (unlikely(size < 8))
+        {
+            if (unlikely(size <= 1))
+            {
+                if (likely(size == 1))
+                {
+                    /// A single byte.
+                    *dst = *src;
+                }
+                /// No bytes remaining.
+            }
+            else if (unlikely(size <= 4))
+            {
+                /// Chunks of 2..3 bytes.
+                details::memcpy_ignore_overlap<2>(dst, src, size);
+            }
+            else
+            {
+                /// Chunks of 4..7 bytes.
+                details::memcpy_ignore_overlap<4>(dst, src, size);
+            }
+        }
+        else if (likely(size <= 16))
+        {
+            /// Chunks of 8..16 bytes.
+            details::memcpy_ignore_overlap<8>(dst, src, size);
+        }
+        else
+        {
+            /// Chunks of 17..32 bytes.
+            details::memcpy_ignore_overlap<16>(dst, src, size);
+        }
+        return ret;
+    }
+
+    constexpr size_t loop_block32_cnt = 8;
+
+    if (unlikely(size > block32_size * 8))
+    {
+        /// Large size with fully unrolled loop.
+        {
+            MEM_CP_HEAD(block32_size);
+
+            // reduce instruction: `offset` = or (`dst`, 0xfffffffffffffff0)
+            auto offset = ssize_t(size_t(dst) % block32_size) - block32_size;
+            dst -= offset;
+            src -= offset;
+            size += offset;
+        }
+
+        /// Aligned unrolled copy.
+        while (size >= block32_size * loop_block32_cnt)
+        {
+#define M(n) details::Block32 c##n = mem_utils::details::load_block32(src + block32_size * (n));
+            M(0)
+            M(1)
+            M(2)
+            M(3)
+            M(4)
+            M(5)
+            M(6)
+            M(7)
+#undef M
+#define M(n) mem_utils::details::write_block32<true>(dst + block32_size * (n), c##n);
+            M(0)
+            M(1)
+            M(2)
+            M(3)
+            M(4)
+            M(5)
+            M(6)
+            M(7)
+#undef M
+            src += block32_size * loop_block32_cnt;
+            dst += block32_size * loop_block32_cnt;
+            size -= block32_size * loop_block32_cnt;
+        }
+        if (size <= block32_size)
+        {
+            MEM_CP_END(block32_size);
+            return ret;
+        }
+    }
+
+    assert(size > block32_size);
+
+    if (unlikely(size <= 2 * block32_size))
```

Comment:
```suggestion
    if (unlikely(size <= (block32_size << 1)))
```

use shift operator to get better performance?

Reply:
Modern compiler itself can optimize such simple case.

Reply:
ok